### PR TITLE
docs: restructure training debugging doc

### PR DESCRIPTION
**Pull request by gh-determined-ai**
## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/training-debug/index.txt [hunk 1]
```diff
@@ -1,343 +1,395 @@
...
-Then create an experiment with the ``--test`` flag (but not the ``--local`` flag):
+#. Launch a notebook or shell as described in :ref:`Step 4 <step4>`.
 
-.. code:: bash
+   On the cluster, testing is the same as :ref:`Step 3 <step3>`, except that the second, model
+   definition argument of the ``det experiment create`` command should be
+   ``/run/determined/workdir``, or ``.``, the current directory if you have not changed the working
+   directory after connecting to the cluster. This is because the ``--context`` specified when
+   creating the shell or notebook is copied to the ``/run/determined/workdir`` directory inside the
+   container, the same as the model definition argument is copied to ``det experiment create``.
```
Comment (rb-determined-ai) [inline docs/training-debug/index.txt:219]: The commas here seem awkward.

> except that the second, model definition argument of the ``det experiment create`` command

Here "second, model definition argument" seems grammatically correct ("second" and "model definition" are two adjectives that both apply) but it feels less clear than just saying the "model definition argument".  I say this because from a CLI perspective, both `experiment` and `create` are technically arguments to the `det` command, but functionally `experiment create` represents a subcommand and a subsubcommand, with arguments after it, so numbering arguments is not crystal clear in this situation.

> should be ``/run/determined/workdir``, or ``.``

No comma after workdir: "should be X or Y".

> or ``.``, the current directory if you

"the current directory" is an appositive, and so needs a comma after it.  But in this case, I would guess it's widely enough known that `.` represents the current directory, and I would just remove the `, the current directory,` appositive entirely.

File: docs/training-debug/index.txt [hunk 2]
```diff
@@ -1,343 +1,395 @@
...
--  The experiment config is fully validated for cluster-managed experiments, more strictly than it
-   is for ``--local --test`` mode. If you get errors related to ``unmarshaling JSON`` when trying to
-   submit the experiment to the cluster, that is an indication that the experiment config has
-   errors. Please review the :ref:`experiment configuration <experiment-configuration>`.
+   .. code:: yaml
 
-Again, if you are unable to identify the root cause of the issue yourself, please do not hesitate to
-contact Determined through our `community support
+      resources:
+      slots_per_trial: 1
```
Comment (rb-determined-ai) [inline docs/training-debug/index.txt:249]: the `slots_per_trial` line needs an indent to be valid yaml

File: docs/training-debug/index.txt [hunk 3]
```diff
@@ -1,343 +1,395 @@
...
+      into a custom Docker image or included in the ``startup-hook.sh`` file in the model definition
+      directory root. See :ref:`startup-hooks` for more information.
+
+   -  Custom bind mounts required in the interactive environment are specified in the experiment
+      configuration.
+
+   -  Environment variables are correctly set in the experiment configuration.
+
+   If no customizations are missing, the following, new layers introduced with a cluster-managed
+   experiment could be the problem cause:
```
Comment (rb-determined-ai) [inline docs/training-debug/index.txt:275]: "could be the cause of the problem" feels far more idiomatic to me than "the problem cause"

File: docs/training-debug/index.txt [hunk 4]
```diff
@@ -1,343 +1,395 @@
...
+   -  For a cluster-based experiment, configured ``checkpoint_storage`` settings are validated
+      before training starts. The message ``Checkpoint storage validation failed``, indicates that
+      you should review the ``checkpoint_storage`` setting values.
+
+   -  The experiment configuration is more strictly validated for cluster-managed experiments than
+      for ``--local --test`` mode. Errors related to ``unmarshaling JSON`` when attempting to submit
+      the experiment to the cluster indicate that the experiment configuration has errors. Review
+      the :ref:`experiment configuration <experiment-configuration>`.
+
+If you are unable to identify the root cause of the issue, contact Determined `community support
```
Comment (rb-determined-ai) [inline docs/training-debug/index.txt:295]: suggest reprasing "the root cause of the issue" to match the phrasing X used before in "If no customizations are missing, the following, new layers introduced with a cluster-managed experiment could be the X"

File: docs/training-debug/index.txt [hunk 5]
```diff
@@ -1,343 +1,395 @@
...
+
+      -  :ref:`master-configuration`
+      -  :ref:`experiment-configuration`
+
+   -  For a cluster-based experiment, configured ``checkpoint_storage`` settings are validated
+      before training starts. The message ``Checkpoint storage validation failed``, indicates that
+      you should review the ``checkpoint_storage`` setting values.
+
+   -  The experiment configuration is more strictly validated for cluster-managed experiments than
+      for ``--local --test`` mode. Errors related to ``unmarshaling JSON`` when attempting to submit
```
Comment (rb-determined-ai) [inline docs/training-debug/index.txt:291]: Oh!  This error message is no longer what we give to users (thankfully, since it was a terrible message we used to show users).

Instead of ``` ``unmarshalling JSON`` ``` we can say ``` ``invalid experiment configuration`` ``` which is the new pattern that would show up in the command output.

File: docs/training-debug/index.txt [hunk 6]
```diff
@@ -1,343 +1,395 @@
...
+   it is set to ``1``. For example:
+
+   .. code:: yaml
+
+      resources:
+        slots_per_trial: 1
+
+#. Create an experiment without the ``--test`` or ``--local`` arguments:
+
+   You might find the ``--follow``, or ``-f``, arguments helpful:
```
Comment (rb-determined-ai) [inline docs/training-debug/index.txt:319]: I do think `flag` is a more correct word than `argument`.  "Flags" generally don't have parameters (like `--verbose`), while "options" usually do (like `--output dump.txt`).  I think of "arguments" as either being positional arguments (neither flags nor options) or as a general term for tokens on the command line (where flags and options and positional arguments are all types of arguments).

In either case, "arguments" should not be plural because it's the same flag, just in a long form and a short form.

File: docs/training-debug/index.txt [hunk 7]
```diff
@@ -1,343 +1,395 @@
...
 
-**How to diagnose failures:** If you are using the ``determined`` library APIs correctly, then
-theoretically distributed training should "just work". However, you should be aware of some common
-pitfalls:
+      -  If you have four Determined agents running with four GPUs each, your ``slots_per_trial``
+         could be ``1``, ``2``, ``3``, or ``4``, which would fit on a single machine.
+      -  A ``slots_per_trial`` value of ``8``, ``12``, or ``16`` would completely utilize a number
+         of agent machines.
+      -  A ``slots_per_trial`` value of ``5``, implies more than one agent but is not a multiple of
+         agent size.
```
Comment (rb-determined-ai) [inline docs/training-debug/index.txt:375]: this is an invalid case, but that is not explained in the text of the bullet

File: docs/training-debug/index.txt [hunk 8]
```diff
@@ -1,343 +1,395 @@
 .. _model-debug:
 
-###########
- Debugging
-###########
+##############
+ Debug Models
```
Comment (rb-determined-ai) [inline docs/training-debug/index.txt:4]: Why not "Debugging Models"?  My thinking is that "Debug Models" does not make it clear that "Debug" is a verb; it could be an adjective the way that it is when you talk about a "debug build" of some C or C++ code.

File: docs/training-debug/index.txt [hunk 9]
```diff
@@ -1,343 +1,395 @@
...
 
--  Your code will conform to the Trial APIs by being organized into a subclass of Determined's
-   ``Trial`` class (indirectly through one of its concrete subclasses, such as
-   :class:`~determined.pytorch.PyTorchTrial`).
+These debugging steps introduce code changes incrementally, working toward a fully functioning
+Determined model. Follow the nine steps as applicable to your environment:
 
--  The code will run in a Docker container on another machine.
+-  Model-related Issues
+      -  `Step 1 - Verify that your code runs locally`_
```
[the remark this answers is not in the document]
Comment (gh-determined-ai) [reply docs/training-debug/index.txt:21]: Referencing the header thusly: href="#step-1-verify-that-your-code-runs-locally"
Pretty standard link encoding.

File: docs/training-debug/index.txt [hunk 10]
```diff
@@ -1,343 +1,395 @@
...
+#. Launch a notebook or shell on the cluster:
 
-**How to test:** First launch a notebook or shell on the cluster, passing the root directory
-containing your model and training scripts to the ``--context`` option on the command line.
+   Pass the root directory containing your model and training scripts in the ``--context`` argument:
 
-Note that changes made to the ``--context`` directory while inside the notebook or shell will not
-affect the original files outside of the notebook or shell; see :ref:`notebook-state` for more
-details.
+   If you are using a Jupyter notebook, enter:
```
[the remark this answers is not in the document]
Comment (gh-determined-ai) [reply docs/training-debug/index.txt:130]: Fixed as requested.

File: docs/training-debug/index.txt [hunk 11]
```diff
@@ -1,343 +1,395 @@
...
+ Step 3 - Verify local test mode
+*********************************
 
-**How to test:** You should create some simple tests to verify that each method of your ``Trial``
-subclass is doing what you think it ought to be doing. There are some simple and practical examples
-of what these tests might look like for :class:`~determined.pytorch.PyTorchTrial` and
-:class:`~determined.keras.TFKerasTrial` in the :meth:`determined.TrialContext.from_config`
-documentation, but ultimately only you know what a reasonable test for your trial needs to look
-like.
+:ref:`Step 2 <step2>`, validated that your Trial API calls work as expected. This step uses your
```
[the remark this answers is not in the document]
Comment (gh-determined-ai) [reply docs/training-debug/index.txt:82]: Fixed as requested.

File: docs/training-debug/index.txt [hunk 12]
```diff
@@ -1,343 +1,395 @@
...
-that it is meeting Determined's requirements.
+      det experiment create myconfig.yaml my_model_dir --local --test
 
-**How to test:** Simply create an experiment, passing ``--local --test`` on the command line:
+   The ``--local`` argument specifies that training occurs where you launched the experiment instead
+   of occurring on a cluster. The ``--test`` argument runs abbreviated workloads to try to detect
+   bugs sooner and exits immediately.
 
-.. code:: bash
+#. The test is considered to have passed if the command completes successfully.
```
[the remark this answers is not in the document]
Comment (gh-determined-ai) [reply docs/training-debug/index.txt:99]: You're right. It belongs as the last sentence of the previous step.

File: docs/training-debug/index.txt (hunk 4 above)
Comment (gh-determined-ai) [reply docs/training-debug/index.txt:295]: Changed according to what I think you meant.

File: docs/training-debug/index.txt (hunk 5 above)
Comment (gh-determined-ai) [reply docs/training-debug/index.txt:291]: Fixed.

File: docs/training-debug/index.txt (hunk 6 above)
Comment (gh-determined-ai) [reply docs/training-debug/index.txt:319]: I'm still leaning toward 'argument' (singular), as in Python's argparse(). 'Option' could also be ok for a CLI. If it was a program, it would be 'parameter'. I think 'flag' is the more general of all of those but usually refers to a programming variable, unofficially.

Let's have one go-around and pick one. They'll know what we mean.

File: docs/training-debug/index.txt (hunk 7 above)
Comment (gh-determined-ai) [reply docs/training-debug/index.txt:375]: Clarified as requested, along with the last bullet for the '32' case.

File: docs/training-debug/index.txt (hunk 8 above)
Comment (gh-determined-ai) [reply docs/training-debug/index.txt:4]: Agree. But headings should not use a gerund (except for Getting Started because it's so ubiquitous). The right title would be 'How to Debug Models' (infinitive form).

File: docs/training-debug/index.txt [hunk 13]
```diff
@@ -1,343 +1,395 @@
...
 
-   det experiment create myconfig.yaml my_model_dir --local --test
+#. Perform failure diagnostics:
```
[the remark this answers is not in the document]
Comment (gh-determined-ai) [reply docs/training-debug/index.txt:101]: Fixed as requested.

File: docs/training-debug/index.txt (hunk 1 above)
Comment (gh-determined-ai) [reply docs/training-debug/index.txt:219]: Fixed. All good suggestions.

File: docs/training-debug/index.txt (hunk 2 above)
Comment (gh-determined-ai) [reply docs/training-debug/index.txt:249]: fixed.

File: docs/training-debug/index.txt (hunk 3 above)
Comment (gh-determined-ai) [reply docs/training-debug/index.txt]: Already fixed.

File: docs/training-debug/index.txt [hunk 14]
```diff
@@ -1,343 +1,397 @@
...
-If you prefer to interact with your model via a Jupyter notebook, try:
+   .. code:: bash
 
-.. code:: bash
+      det notebook start --context my_model_dir
+      # Your browser should automatically open the notebook.
 
-   det notebook start --context my_model_dir
-   # Your browser should automatically open the notebook.
+   If you are using SSH to interact with your model, enter:
```
Comment (rb-determined-ai) [inline docs/training-debug/index.txt:137]: "if you prefer"...

File: docs/training-debug/index.txt [hunk 15]
```diff
@@ -1,343 +1,397 @@
...
+This step is the same as :ref:`Step 2 <step2>`, except the original code runs on the Determined
+cluster instead of locally.
 
-**How to diagnose failures:** Failure diagnosis is a combination of the failure diagnosis for Step 2
-and for Step 4.
+#. Launch a notebook or shell:
 
-6. Does local test mode work in a notebook or shell?
-====================================================
+   If you are using a Jupyter notebook, enter:
```
Comment (rb-determined-ai) [inline docs/training-debug/index.txt:183]: "if you prefer"

File: docs/training-debug/index.txt [hunk 16]
```diff
@@ -1,343 +1,397 @@
...
 
-**How to test:** Launch a notebook or shell as in Step 4.
+      det notebook start --context my_model_dir
+      # Your browser should automatically open the notebook.
 
-Once you are on the cluster, testing is identical to Step 3, with the important caveat that the
-model definition argument to ``det experiment create`` (the second positional argument) should
-always be ``/run/determined/workdir`` (or ``.`` if you have not changed the working directory from
-when you first connected to the cluster).
+   If you are using SSH to interact with your model, enter:
```
Comment (rb-determined-ai) [inline docs/training-debug/index.txt:190]: "if you prefer"

File: docs/training-debug/index.txt [hunk 17]
```diff
@@ -1,343 +1,397 @@
...
-**How to diagnose failures:**
+   After you are on the cluster, testing is the same as :ref:`Step 1 <step1>`.
 
--  If you are unable to start the container with a message about the context directory exceeding the
-   maximum allowed size, it is because the ``--context`` directory cannot be more than 95MB in size.
-   If you need files larger than that as part of your model definition, consider setting up a bind
-   mount via the ``bind_mounts`` field of the :ref:`task configuration
-   <command-notebook-configuration>`. The :ref:`prepare-data` document lists some additional
-   strategies for accessing files inside a containerized environment that may also be relevant.
+#. Perform the following failure diagnostic steps:
```
Comment (rb-determined-ai) [inline docs/training-debug/index.txt:152]: "Diagnose failures", maybe?

File: docs/training-debug/index.txt [hunk 18]
```diff
@@ -1,343 +1,397 @@
...
-.. code:: bash
+   On the cluster, testing is the same as :ref:`Step 3 <step3>`, except that the second model
+   definition argument of the ``det experiment create`` command should be
+   ``/run/determined/workdir`` or ``.`` if you have not changed the working directory after
+   connecting to the cluster. This is because the ``--context`` specified when creating the shell or
+   notebook is copied to the ``/run/determined/workdir`` directory inside the container, the same as
+   the model definition argument is copied to ``det experiment create``.
 
-   det experiment create myconfig.yaml my_model_dir --test
+#. Perform failure diagnosis:
```
Comment (rb-determined-ai) [inline docs/training-debug/index.txt:221]: "Diagnose failures"

File: docs/training-debug/index.txt [hunk 19]
```diff
@@ -1,343 +1,397 @@
...
 
-**How to diagnose failures:** If you were able to run local test mode inside a notebook or shell,
-but you are unable to successfully submit an experiment, you should focus on making sure that any
-customizations you made to get it to work in the notebook or shell have been properly replicated in
-your :ref:`experiment config <experiment-configuration>`:
+   Follow the combined failure diagnostic steps used for :ref:`Step 3 <step3>` and :ref:`Step 4
```
Comment (rb-determined-ai) [inline docs/training-debug/index.txt:223]: suggested reword: "Follow the combined failure diagnoses from..."

Similar motivation as "Diagnose failures" vs "Perform failure diagnosis".  I don't feel strongly about this reword at all.

File: docs/training-debug/index.txt (hunk 4 above)
Comment (rb-determined-ai) [reply docs/training-debug/index.txt]: What I mean is: a couple of sentences earlier you say 

> If no customizations are missing, the following new layers introduced with a cluster-managed experiment could be the cause of the problem:"

I'm suggesting changing what you had before: 

> If you are unable to identify ~the root cause of the issue~, ...

to

> If you are unable to identify **the cause of the problem**, ...

specifically to match the phrasing you used just before.

I'm not a big fan of the newest phrasing:

> If no customizations are missing and the new layers introduced with a cluster-managed experiment are not the cause of the problem, contact Determined

because I would always rather encourage users to try to figure it out themselves before coming to us, and also because a user might find it very difficult to determine if "the new layers introduced with a cluster-managed experiment are not the cause of the problem"

Comment (gh-determined-ai) [reply docs/training-debug/index.txt]: Agree. Fixed.

File: docs/training-debug/index.txt (hunk 14 above)
Comment (gh-determined-ai) [reply docs/training-debug/index.txt:137]: fixed

File: docs/training-debug/index.txt (hunk 15 above)
Comment (gh-determined-ai) [reply docs/training-debug/index.txt:183]: fixed.

File: docs/training-debug/index.txt (hunk 16 above)
Comment (gh-determined-ai) [reply docs/training-debug/index.txt:190]: fixed.

File: docs/training-debug/index.txt (hunk 19 above)
Comment (gh-determined-ai) [reply docs/training-debug/index.txt:223]: Fixed.

**Review by rb-determined-ai — APPROVED** [review]
This is great!  It looks like professional documentation now : )
